core/common/src/main/java/alluxio/collections/FieldIndex.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.collections;
+
+import java.util.Set;
+
+/**
+ * An interface representing an index for this {@link IndexedSet}, each index for this set must
+ * implement the interface to define how to get the value of the field chosen as the index. Users
+ * must use the same instance of the implementation of this interface as the parameter in all
+ * methods of {@link IndexedSet} to represent the same index.
+ *
+ * @param <T> type of objects in this {@link IndexedSet}
+ */
+interface FieldIndex<T> {
+  /**
+   * Gets the value of the field that serves as index.
+   *
+   * @param o the instance to get the field value from
+   * @return the field value, which is just an Object
+   */
+  Object getFieldValue(T o);
```

Comment:
An index is a data structure used by IndexedSet to look up elements by certain fields in constant time. The index needs to know how to get a field value out of the element, but there's no reason that it should make that information public. Whoever created the index must already know how to get a field value out of an element, or they wouldn't have been able to create the index in the first place.

If you still think this method belongs here, can you give an example of when it would be useful?